sw/otbn/crypto/p256_isoncurve.s

```diff
@@ -84,3 +94,120 @@ p256_isoncurve:
   jal       x1, mul_modp
 
   ret
+
+/**
+ * Check if a provided public key is valid.
+ *
+ * For a given public key (x, y), check that:
+ * - x and y are both fully reduced mod p
+ * - (x, y) is on the P-256 curve.
+ *
+ * Note that, because the point is in affine form, it is not possible that (x,
+ * y) is the point at infinity. In some other forms such as projective
+ * coordinates, we would need to check for this also.
+ *
+ * This routine sets `ok` to false if the check fails and immediately exits the
+ * program. If the check succeeds, `ok` is unmodified.
+ *
+ * @param[in] dmem[x]: Public key x-coordinate.
+ * @param[in] dmem[y]: Public key y-coordinate.
+ * @param[out] dmem[ok]: success/failure of basic checks (32 bits)
+ *
+ * clobbered registers: x2, x3, x19, x20, w0, w2, w19 to w29
+ * clobbered flag groups: FG0
+ */
+p256_check_public_key:
+  /* Init all-zero register. */
+  bn.xor   w31, w31, w31
+
+  /* Load domain parameter p.
+       w29 <= dmem[p256_p] = p */
+  li        x2, 29
+  la        x3, p256_p
+  bn.lid    x2, 0(x3)
+
+  /* Load public key x-coordinate.
+       w2 <= dmem[x] = x */
+  li        x2, 2
+  la        x3, x
+  bn.lid    x2, 0(x3)
+
+  /* Compare x to p.
+       FG0.C <= (x < p) */
+  bn.cmp    w2, w29
+
+  /* Fail if FG0.C is false. */
+  csrrs     x2, FG0, x0
+  andi      x2, x2, 1
+  bne       x2, x0, _x_valid
+  jal       x0, p256_invalid_input
+  unimp
+
+  _x_valid:
+
+  /* Load public key y-coordinate.
+       w2 <= dmem[y] = y */
+  li        x2, 2
+  la        x3, y
+  bn.lid    x2, 0(x3)
+
+  /* Compare y to p.
+       FG0.C <= (y < p) */
+  bn.cmp    w2, w29
+
+  /* Fail if FG0.C is false. */
+  csrrs     x2, FG0, x0
+  andi      x2, x2, 1
+  bne       x2, x0, _y_valid
+  jal       x0, p256_invalid_input
+  unimp
+
+  _y_valid:
+
+  /* Compute both sides of the Weierstrauss equation.
+       w18 <= (x^3 + ax + b) mod p
+       w19 <= (y^2) mod p */
+  jal      x1, p256_isoncurve
+
+  /* Compare the two sides of the equation.
+       FG0.Z <= (y^2) mod p == (x^2 + ax + b) mod p */
+  bn.cmp    w18, w19
+
+  /* Fail if FG0.Z is false; otherwise return. */
+  csrrs     x2, FG0, x0
+  andi      x2, x2, 8
+  bne       x2, x0, _xy_on_curve
```

Comment:
Now that `invalid input` is now an actual handler, rather than just invoking `unimp`. I think it may be possible to simplify this. How about:
```
/* jump to invalid input if FG0.Z is false */
be x2, x0, p256_invalid_input
ret
```
This would allow us to get rid of `_xy_on_curve`.

Reply:
As you said yesterday, in this case what's simplest to code is also most vulnerable to fault injection.  For FI reasons, I always try to write sensitive branches with `bne`, so if it's not the _exact value_ that I expect then the "bad" branch gets taken. Otherwise if it was `beq`, an attacker could inject any 1-bit fault into x2 and get into the "good" branch.

Reply:
After I wrote this I was beginning to think it was a FI thing :)